src/webgpu/shader/validation/expression/call/builtin/refract.spec.ts

```diff
@@ -0,0 +1,164 @@
+      // If the k is near zero it may fail on some implementations which implement sqrt as
+      // 1/inversesqrt, so skip the test.
+      if (vCheck.isNearZero(k)) {
+        t.skip(`K value, ${k}, is at or near 0.`);
+      }
+
+      const ca = vCheck.checkedResult(c * a);
+      const cbda = vCheck.checkedResult(c * b_dot_a);
+      const sqrt_k = vCheck.checkedResult(Math.sqrt(k));
+      const cdba_sqrt_k = vCheck.checkedResult(cbda + sqrt_k);
+      const cdba_sqrt_k_b = vCheck.checkedResult(cdba_sqrt_k * b);
+      vCheck.checkedResult(ca - cdba_sqrt_k_b);
+    }
+
+    // Validates refract(vecN(a), vecN(b), c);
+    validateConstOrOverrideBuiltinEval(
+      t,
+      builtin,
+      vCheck.allChecksPassed(),
+      [type.create(t.params.a), type.create(t.params.b), scalarType.create(t.params.c)],
+      t.params.stage
+    );
+  });
+
+const kArgCases = {
+  good: '(vec3(0), vec3(1), 2.0)',
+  bad_no_parens: '',
+  // Bad number of args
+  bad_0args: '()',
+  bad_1arg: '(vec3(0))',
+  bad_2arg: '(vec3(0), vec3(1))',
+  bad_3arg: '(vec3(0), vec3(1), 2.0, vec3(3))',
+  // Mismatched vec sizes
+  bad_vec2_vec3: '(vec2(0), vec3(1), 2.0)',
+  bad_vec3_vec4: '(vec3(0), vec4(1), 2.0)',
+  bad_vec4_vec2: '(vec4(0), vec2(1), 2.0)',
+  // Bad value for arg 0
+  bad_0bool: '(false, vec3(1), 2.0)',
+  bad_0array: '(array(1.1,2.2), vec3(1), 2.0)',
+  bad_0struct: '(modf(2.2), vec3(1), 2.0)',
+  bad_0int: '(0i, vec3(1), 2.0)',
+  bad_0uint: '(0u, vec3(1), 2.0)',
+  bad_0f32: '(0.0, vec3(1), 2.0)',
+  bad_0f16: '(0.0h, vec3(1), 2.0)',
+  bad_0veci: '(vec3i(0), vec3(1), 2.0)',
+  bad_0vecu: '(vec3u(0), vec3(1), 2.0)',
+  // Bad value type for arg 1
+  bad_1bool: '(vec3(0), true, 2.0)',
+  bad_1array: '(vec3(0), array(1.1,2.2), 2.0)',
+  bad_1struct: '(vec3(0), modf(2.2), 2.0)',
+  bad_1int: '(vec3(0), 1i, 2.0)',
+  bad_1uint: '(vec3(0), 1u, 2.0)',
+  bad_1f32: '(vec3(0), 1.0, 2.0)',
+  bad_1f16: '(vec3(0), 1.0h, 2.0)',
+  bad_1veci: '(vec3(0), vec3i(1), 2.0)',
+  bad_1vecu: '(vec3(0), vec3u(1), 2.0)',
+  // Bad value type for arg 2
+  bad_2bool: '(vec3(0), vec3(1), true)',
+  bad_2array: '(vec3(0), vec3(1), array(1.1,2.2))',
+  bad_2struct: '(vec3(0), vec3(1), modf(2.2))',
+  bad_2int: '(vec3(0), vec3(1), 2i)',
+  bad_2uint: '(vec3(0), vec3(1), 2u)',
+  bad_2veci: '(vec3(0), vec3(1), vec3i(2))',
+  bad_2vecu: '(vec3(0), vec3(1), vec3u(2))',
+  bad_2vecf: '(vec3(0), vec3(1), vec3f(2))',
+  bad_2vech: '(vec3(0), vec3(1), vec3h(2))',
+};
+
+g.test('args')
+  .desc(`Test compilation failure of ${builtin} with variously shaped and typed arguments`)
+  .params(u => u.combine('arg', keysOf(kArgCases)))
+  .fn(t => {
+    t.expectCompileResult(
+      t.params.arg === 'good',
+      `const c = ${builtin}${kArgCases[t.params.arg]};`
+    );
+  });
+
+const kReturnType = 'vec3f';
+
+g.test('return')
+  .desc(`Test ${builtin} return value type`)
+  .params(u => u.combine('type', ['vec3f', 'vec3u', 'vec3i', 'vec2f', 'u32', 'i32', 'f32', 'bool']))
```

Comment:
This is slightly weak sauce because only one one of the 9 possible return types is checked.
But I'm ok with that: those other ones are surely(?) checked in the execution tests, at least implicitly.

Reply:
I see what you mean. Refactored to do a better job of it, I'd appreciate you taking another look!